### ISSUE TITLE
Issues with the objc `named_framework_to_proto_path_mappings_path` parameter

### ISSUE THREAD

**Issue by justinseanmartin** [google/protobuf#1814]
FYI - @thomasvl (looks like you implemented it?)

We have a problem attempting to use the framework support in the ObjectiveC compiler. The issue is two-fold:
1. Local proto dependencies within the same framework use "path/to/The.pbobjc.h" dependencies, but the framework imports expect them to be flattened as <framework/The.pbobjc.h> (or paths to be searched recursively).
2. We may have multiple protos with the same name in the same module under different folder paths, which will conflict on being flattened. e.g. squareup/group1/Options.pbobjc.h & squareup/group2/Options.pbobjc.h (no naming conflict by having a unique prefix)

For this reason, I believe we should be maintaining the folder structure within the frameworks and the imports that are added should continue to use the explicit full path to the dependency. I believe the fix would look like:

```
diff --git a/src/google/protobuf/compiler/objectivec/objectivec_file.cc b/src/google/protobuf/compiler/objectivec/objectivec_file.cc
index cf8c34e..421d4e1 100644
--- a/src/google/protobuf/compiler/objectivec/objectivec_file.cc
+++ b/src/google/protobuf/compiler/objectivec/objectivec_file.cc
@@ -110,14 +110,14 @@ void ImportWriter::AddFile(const FileGenerator* file) {
   if (proto_lookup != proto_file_to_framework_name_.end()) {
     other_framework_imports_.push_back(
         proto_lookup->second + "/" +
-        FilePathBasename(file_descriptor) + extension);
+        file->Path() + extension);
     return;
   }

   if (!options_.generate_for_named_framework.empty()) {
     other_framework_imports_.push_back(
         options_.generate_for_named_framework + "/" +
-        FilePathBasename(file_descriptor) + extension);
+        file->Path() + extension);
     return;
   }

```

This leads to imports looking like:

```
#import <Framework/path/to/The.pbobjc.h>
```

I'm not sure if there is something I'm missing though. I'd be happy to open a PR with this change if this is an agreeable solution. Thanks for your feedback.

CC: @dnkoutso

**Comment by justinseanmartin** [comment]
There appears to be another problem with framework support. In the imports that are generated for source files, the headers from within the same module need to be imported using quoted strings and relative paths rather than as framework imports. I'm working on a fix for this now.

**Comment by sergiocampama** [comment]
One thing you can do is add the path to the mappings file, as so:

```
Path/To/My/Framework:my_proto.proto,other_proto.proto
```

Does this solve your use case?

**Comment by justinseanmartin** [comment]
I don't think it does fully solve the issue. I could set it up as:

Framework/path/to:my_proto.proto
Framework/other/path/to:other_proto.proto

That part is totally feasible. The problem is it doesn't differentiate between headers in the current framework and headers pulled in from other frameworks. Headers pulled in for the current framework need to use "" relative imports, while pulling in headers from other frameworks need to use <> absolute framework imports. 

Cocoapods already handles retaining the directory structure when using the `header_mappings_dir` for Frameworks. This is necessary when we have protos with the same name under different namespaces. For example, we may have "service1/api.proto" and "service2/api.proto". Otherwise, we'd end up with collisions when the folder structure is flattened and have to Api.pbobjc.[hm] files.

**Comment by sergiocampama** [comment]
so what you're saying is basically:

```
FrameworkA:a.proto,b.proto
FrameworkB:c.proto,d.proto
```

which would generate:

```
// A.pbobjc.m
#import "path/to/A.pbobjc.h"
#import "path/to/B.pbobjc.h"
#import <FrameworkB/C.pbobjc.h>
#import <FrameworkB/D.pbobjc.h>
```
...
```
// C.pbobjc.m
#import <FrameworkA/A.pbobjc.h>
#import <FrameworkA/B.pbobjc.h>
#import "path/to/C.pbobjc.h"
#import "path/to/D.pbobjc.h"
```
?

**Comment by justinseanmartin** [comment]
Close:
```
FrameworkA:path1/a.proto,path2/b.proto
FrameworkB:c.proto,d.proto
```

would generate:
```
// A.pbobjc.m
#import <FrameworkB/C.pbobjc.h>
#import <FrameworkB/D.pbobjc.h>
#import "A.pbobjc.h"
#import "../B.pbobjc.h"
```

The difference is that quoted paths are relative from the current source file. At least, this is what is working for us internally with latest Protobuf w/ `named_framework_to_proto_path_mappings_path` & Cocoapods w/ `use_frameworks!`. I think this should generically work for everyone though. I'll push a branch with my changes for inspection.

**Comment by sergiocampama** [comment]
actually, would it be possible for you to upload a sample project that mimics your setup? including cocoapods and how you generate the protos?

**Comment by justinseanmartin** [comment]
Here is the change itself that we're using at the moment:
https://github.com/justinseanmartin/protobuf/commit/0f6ae48fe3e0836675b747e00617965d766f5525

Our protobuf generation is decoupled from the app build process and wouldn't be easy to create a sample. I can probably do it, but it might take some time.

**Comment by sergiocampama** [comment]
Is renaming the proto files possible? Would that be a sensible workaround?

**Comment by justinseanmartin** [comment]
You mean renaming proto files to ensure there is no naming conflict when the protos are flattened? We are doing something similar to ensure unique class names (adding an objc prefix dynamically based on the folder structure).

That said, I don't know if flattening the folder structure would help with this problem or not. We'd still end up with referencing the headers for the current library as `<FrameworkA/a.pbobjc.h>` from `a.pbobjc.m`, which I think is generally considered incorrect.

The changes in my branch seem like something that should be supported by the protoc compiler to enable this. It works for us when integrated as both framework and static library targets, so I'd assume the same for other project configurations as well.

**Comment by sergiocampama** [comment]
For the class name collision, you could use the expected_prefixes_path option to pass a list of proto packages and their required prefix, instead of dynamically adding a prefix based on path. See the next link for information on how to use it...

https://github.com/google/protobuf/blob/master/src/google/protobuf/compiler/objectivec/objectivec_generator.cc#L65

Let me get more familiar with how cocoapods works with frameworks, as I've never had the pleasure...

**Comment by thomasvl** [comment]
At one point, it seems like cocoapods was having some real issues with path preserving in frameworks?  Is that fully resolved in the 1.0.x world?  (we might also have been just hitting issues because the protobuf projects uses "src/" and we don't really want that in the framework hierarchy also (for the Protobuf's pod).

**Comment by justinseanmartin** [comment]
Yea, it seems to be addressed in the 1.0.0 or 1.0.1 releases. Xcode project settings don't make it easy to preserve paths in frameworks, but cocoapods as a copy resources step for each individual folder. I think it was broken before that, but I'm not certain.
